src/semidiscretization/semidiscretization_euler_acoustics.jl

```diff
@@ -0,0 +1,201 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+
+
+"""
+    SemidiscretizationEulerAcoustics(semi_acoustics::SemiAcoustics, semi_euler::SemiEuler;
+                                     source_region=x->true, weights=x->1.0)
+
+!!! warning "Experimental code"
+    This semidiscretization is experimental and may change in any future release.
+
+Construct a semidiscretization of the acoustic perturbation equations that is coupled with
+the compressible Euler equations via source terms. Both semidiscretizations have to use the same
+mesh and solvers with a shared basis. The coupling region is described by a function `source_region`
+that maps the coordinates of a single node to `true` or `false` depending on whether the point lies
+within the coupling region or not. A weighting function `weights` that maps coordinates to weights
+is applied to the acoustic source terms.
+Note that this semidiscretization should be used in conjunction with
+[`EulerAcousticsCouplingCallback`](@ref) and only works in two dimensions.
+"""
+struct SemidiscretizationEulerAcoustics{SemiAcoustics, SemiEuler, Cache} <: AbstractSemidiscretization
+  semi_acoustics::SemiAcoustics
+  semi_euler::SemiEuler
+  performance_counter::PerformanceCounter
+  cache::Cache
+
+  function SemidiscretizationEulerAcoustics{SemiAcoustics, SemiEuler, Cache}(
+    semi_acoustics, semi_euler, cache) where {SemiAcoustics, SemiEuler, Cache}
+
+    # Currently both semidiscretizations need to use a shared mesh
+    @assert semi_acoustics.mesh == semi_euler.mesh
+
+    # Check if both solvers use the same polynomial basis
+    @assert typeof(semi_acoustics.solver.basis) == typeof(semi_euler.solver.basis)
+    for field in fieldnames(typeof(semi_acoustics.solver.basis))
+      value_acoustics = getfield(semi_acoustics.solver.basis, field)
+      value_euler = getfield(semi_euler.solver.basis, field)
+      @assert value_acoustics == value_euler
+    end
```

Comment:
Why `typeof` etc. instead of just checking equality of the bases?

Reply:
Probably because of this:
```julia
julia> using Trixi

julia> b1 = LobattoLegendreBasis(3)
LobattoLegendreBasis{Float64} with polynomials of degree 3

julia> b2 = LobattoLegendreBasis(3)
LobattoLegendreBasis{Float64} with polynomials of degree 3

julia> b1 == b2
false
```

Reply:
> Probably because of this:
> ...

Exactly.

Reply:
Then it looks like we should implement `==(b1::LobattoLegendreBasis, b2:;LobattoLegendreBasis)` appropriately, shouldn't we?

Reply:
Can you tell why the default comparison does not work? Also, I have never implemented a comparison operator for a custom type like this before in Julia - do you have any pointers as to where to look for an example?

Reply:
`==` defaults to `===`. We use `Matrix` fields in `LobattoLegendreBasis`. Thus, two bases may store matrices with the same content (checked by `==`) but not the same pointer (checked by `===`).

Reply:
```julia
help?> ==
search: == === !==

  ==(x, y)

  Generic equality operator. Falls back to ===. Should be implemented for all types with a notion of equality,
  based on the abstract value that an instance represents. For example, all numeric types are compared by
  numeric value, ignoring type. Strings are compared as sequences of characters, ignoring encoding. For
  collections, == is generally called recursively on all contents, though other properties (like the shape for
  arrays) may also be taken into account.
```

Reply:
Thus, basically the same recursive logic as the one used here is recommended.

Reply:
@ranocha Thanks for the pointers!
@lchristm could you please try to implement such an equality operator for the `LobattoLegendreBasis` type and use your loop above to do the actual comparisons? Then, you can use the `==` in the assert.

Reply:
Implemented the equality operator in 104fccf.